index.js

```diff
@@ -0,0 +1,91 @@
+import express from 'express';
+import fs from 'fs';
+
+const server = express();
+
+let requestExpirationDates = [];
+let nextExpirationDate = new Date();
+let nextComparsionInterval = 1;
```

Comment:
Global variables are convenient, as you can access them from anywhere. In this small context this is a valid solution, but it also makes it harder to understand data flows. In bigger setups they can lead to bugs, which are hard to track down, that's why we usually avoid them.

Reply:
Do you have an idea, how to implement this without global variables?

Reply:
Hi Andy, thank you for your comments and feedbacks. 

I can think of two ways to implement this without using global variables, one way that I think is bad and the other that I think is good.

The first - the worst one - would be storing the values of these variables in a file. But this way would not be nice because we would have to access the file whenever we need to read or update the values of these variables, so it would cost performance.

The other way - the good one in my opinion - would be modularizing the code so that we have a defined scope for these variables, in this case I thought of a class. I made some adjustments to the code to test my idea and basically what I did was create a class that would be the service (where the request counter business rules would be and where these variables would be manipulated) and keep the index.js just to call some methods of this class and to have the controllers methods. 

The `index.js` will be like:

```
import express from 'express';
import { RequestCounterService } from './service.js';

const server = express();

const requestCounterService = new RequestCounterService();

requestCounterService.startTimer();

requestCounterService.readFileAndUpdateNumberOfRequests();

server.get('/requestCounter', (req, res) =>{

  requestCounterService.updateRequestExpirationDates();

  requestCounterService.getRequests().then(
    (val) => { 
      return res.json(val);
    }
  )
});

server.listen(3000);
```

And the service class will have another methods:

```
import fs from 'fs';

export class RequestCounterService {
    
    constructor() {
        this.requestExpirationDates = [];
        this.nextExpirationDate = new Date();
        this.nextComparsionInterval = 1;
    }    

    startTimer() {
        setInterval(() => { 
            if(this.nextExpirationDate <= new Date() && this.requestExpirationDates.length !== 0){
                this.requestExpirationDates.every(element => {
                if(element <= this.nextExpirationDate){
                  this.requestExpirationDates.shift();
                  this.updateNextExpirationDateAndComparsionInterval();
                }
                else
                  return false;
              });
              this.writeOnFile();
            }
        
        }, this.nextComparsionInterval);
    }

    getRequests() {
        return new Promise((resolve) => {
          this.writeOnFile();
          resolve(this.requestExpirationDates.length);
        })
    }
      
    writeOnFile() {
        fs.writeFile('./assets/requests.txt', this.requestExpirationDates.toString(), (err) => {
          if (err) throw new Error('Error writing to file.');
        });
    }
      
    readFileAndUpdateNumberOfRequests() {
        fs.readFile('./assets/requests.txt', 'utf8', (err, data) => {
          if(err){
            if(err.code == 'ENOENT' )
              this.requestExpirationDates = [];      
            else        
              throw new Error('Error reading file.');      
          }
          else{
            this.updateRequestDatesWithFileContent(data);  
            this.updateNextExpirationDateAndComparsionInterval();
          }
        });
    }
      
    updateRequestDatesWithFileContent(data){
        let dataArray = data.toString().split(",");
            dataArray.forEach(element => {
              if(this.isValidDate(new Date(element)))
                this.requestExpirationDates.push(new Date(element));        
            });      
    }
      
    isValidDate(date) {
        return !isNaN(date);
    }
      
    updateNextExpirationDateAndComparsionInterval() {
        this.nextExpirationDate = this.requestExpirationDates.length === 0 ? new Date() : this.requestExpirationDates[0];
        this.nextComparsionInterval = this.requestExpirationDates.length === 0 ? 1000*60 : this.nextExpirationDate - new Date();
    }

    updateRequestExpirationDates() {
        let expirationDate = new Date(new Date().getTime() + 1000 * 60);
        if(this.requestExpirationDates.length === 0){
            this.nextExpirationDate = expirationDate;
        }
        this.requestExpirationDates.push(expirationDate);
    }
}
```

I can't say if this is the best approach, I had to research a little bit how modularization works in node so I don't know if I used all the syntaxes correctly, but that would be more or less what I would do in other languages.

I would love to hear your feedback on my suggestions for solving this issue and also how you would solve it if you don't mind sharing it with me. So I can learn one more thing about node.js.